src/plone/restapi/services/users/update.py

```diff
@@ -47,29 +55,78 @@ def reply(self):
             alsoProvides(self.request,
                          plone.protect.interfaces.IDisableCSRFProtection)
 
-        for key, value in user_settings_to_update.items():
-            if key == 'password':
-                user.userSetPassword(user.getUserId(), value)
-            elif key == 'username':
-                set_own_login_name(user, value)
-            else:
-                user.setMemberProperties(mapping={key: value})
+        security = getAdapter(self.context, ISecuritySchema)
+
+        if self.can_manage_users:
+            for key, value in user_settings_to_update.items():
+                if key == 'password':
+                    user.userSetPassword(user.getUserId(), value)
+                elif key == 'username':
+                    set_own_login_name(user, value)
+                else:
+                    user.setMemberProperties(mapping={key: value})
 
-        roles = user_settings_to_update.get('roles', {})
-        if roles:
-            to_add = [key for key, enabled in roles.items() if enabled]
-            to_remove = [key for key, enabled in roles.items() if not enabled]
+            roles = user_settings_to_update.get('roles', {})
+            if roles:
+                to_add = [key for key, enabled in roles.items() if enabled]
+                to_remove = [key for key, enabled in roles.items()
+                             if not enabled]
 
-            target_roles = set(user.getRoles()) - set(to_remove)
-            target_roles = target_roles | set(to_add)
+                target_roles = set(user.getRoles()) - set(to_remove)
+                target_roles = target_roles | set(to_add)
 
-            acl_users = getToolByName(self.context, 'acl_users')
-            acl_users.userFolderEditUser(
-                principal_id=user.id,
-                password=None,
-                roles=target_roles,
-                domains=user.getDomains(),
-            )
+                acl_users = getToolByName(self.context, 'acl_users')
+                acl_users.userFolderEditUser(
+                    principal_id=user.id,
+                    password=None,
+                    roles=target_roles,
+                    domains=user.getDomains(),
+                )
+        elif self._get_current_user == self._get_user_id:
+            for key, value in user_settings_to_update.items():
+                if key == 'password' and \
+                   security.enable_user_pwd_choice and \
+                   self.can_set_own_password:
+                    user.userSetPassword(user.getUserId(), value)
```

Comment:
whitespace

Reply:
I'm not to fond of the line wrapping, but I don't see an easy way out.
We could split this up in separate functions outside the class and then call them?

Reply:
Following clean code best practices would mean to split that up into functions or methods. Though, I don't see how this would become significantly "cleaner". If somebody wants to put efforts into that, fine. If not I am also ok with merging. Working code is better than no code at all. ;)